host/linux/posix.c

```diff
@@ -0,0 +1,1133 @@
+
+    return read((int)fd, buf, count);
+}
+
+ssize_t oe_posix_write_ocall(oe_host_fd_t fd, const void* buf, size_t count)
+{
+    errno = 0;
+
+    return write((int)fd, buf, count);
+}
+
+oe_off_t oe_posix_lseek_ocall(oe_host_fd_t fd, oe_off_t offset, int whence)
+{
+    errno = 0;
+
+    return lseek((int)fd, offset, whence);
+}
+
+int oe_posix_close_ocall(oe_host_fd_t fd)
+{
+    errno = 0;
+
+    return close((int)fd);
+}
+
+oe_host_fd_t oe_posix_dup_ocall(oe_host_fd_t oldfd)
+{
+    errno = 0;
+
+    return dup((int)oldfd);
+}
+
+uint64_t oe_posix_opendir_ocall(const char* name)
+{
+    return (uint64_t)opendir(name);
+}
+
+int oe_posix_readdir_ocall(
+    uint64_t dirp,
+    uint64_t* d_ino,
+    int64_t* d_off,
+    uint16_t* d_reclen,
+    uint8_t* d_type,
+    char* d_name,
+    size_t d_namelen)
+{
+    int ret = -1;
+    struct dirent* ent;
+
+    errno = 0;
+
+    if (!dirp)
+    {
+        errno = EBADF;
+        goto done;
+    }
+
+    if (!d_ino || !d_off || !d_reclen || !d_type || !d_name)
+    {
+        errno = EINVAL;
+        goto done;
+    }
+
+    errno = 0;
+
+    if (!(ent = readdir((DIR*)dirp)))
+    {
+        if (errno)
+            goto done;
+
+        ret = -1;
+        goto done;
+    }
+
+    {
+        size_t len = strlen(ent->d_name);
+
+        *d_ino = ent->d_ino;
+        *d_off = ent->d_off;
+        *d_reclen = ent->d_reclen;
+        *d_type = ent->d_type;
+
+        if (len >= d_namelen)
+        {
+            errno = ENAMETOOLONG;
+            goto done;
+        }
+
+        memcpy(d_name, ent->d_name, len + 1);
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+void oe_posix_rewinddir_ocall(uint64_t dirp)
+{
+    if (dirp)
+        rewinddir((DIR*)dirp);
+}
+
+int oe_posix_closedir_ocall(uint64_t dirp)
+{
+    errno = 0;
+
+    return closedir((DIR*)dirp);
+}
+
+int oe_posix_stat_ocall(const char* pathname, struct oe_stat* buf)
+{
+    int ret = -1;
+    struct stat st;
+
+    errno = 0;
+
+    if (!buf)
+        goto done;
+
+    if ((ret = stat(pathname, &st)) == -1)
+        goto done;
+
+    buf->st_dev = st.st_dev;
+    buf->st_dev = st.st_dev;
+    buf->st_ino = st.st_ino;
+    buf->st_nlink = st.st_nlink;
+    buf->st_mode = st.st_mode;
+    buf->st_uid = st.st_uid;
+    buf->st_gid = st.st_gid;
+    buf->st_rdev = st.st_rdev;
+    buf->st_size = st.st_size;
+    buf->st_blksize = st.st_blksize;
+    buf->st_blocks = st.st_blocks;
+    buf->st_atim.tv_sec = st.st_atim.tv_sec;
+    buf->st_atim.tv_nsec = st.st_atim.tv_nsec;
+    buf->st_mtim.tv_sec = st.st_mtim.tv_sec;
+    buf->st_mtim.tv_nsec = st.st_mtim.tv_nsec;
+    buf->st_ctim.tv_sec = st.st_ctim.tv_sec;
+    buf->st_ctim.tv_nsec = st.st_ctim.tv_nsec;
+
+done:
+    return ret;
+}
+
+int oe_posix_access_ocall(const char* pathname, int mode)
+{
+    errno = 0;
+
+    return access(pathname, mode);
+}
+
+int oe_posix_link_ocall(const char* oldpath, const char* newpath)
+{
+    errno = 0;
+
+    return link(oldpath, newpath);
```

Comment:
@gupta-ak Can edlgen be used to autogen these forwarding functions? Not saying that w should right now, just curious.
CC: @andschwa for future ideas. Another team has expressed interest in having forwarding function auto-generated as well.

Reply:
Interesting future idea. :)

Reply:
Yeah, it can generate the C calls on the enclave+host side and the edl as well. I made it also do things like make it work properly on `accept`, so like `accept(int, struct sockaddr*, socklen_t*) -> accept(int, [in,out, size=socklen_in] struct sockaddr*, [in, out] socklen_t*, size_t socklen_in)`.

Since it was a prototype, I didn't make it templatized, so the generated code looks different, but if there is an ask from another team, we could add to it and make it a proper tool.